rskj-core/src/main/java/co/rsk/trie/Trie.java

```diff
@@ -766,26 +766,26 @@ private Trie put(TrieKeySlice key, byte[] value, boolean isRecursiveDelete) {
             return trie;
         }
 
-        Optional<Trie> leftOpt = trie.left.getNode();
-        Optional<Trie> rightOpt = trie.right.getNode();
-        if (leftOpt.isPresent() && rightOpt.isPresent()) {
-            return trie;
-        }
-
-        if (!leftOpt.isPresent() && !rightOpt.isPresent()) {
+        // both left and right exist (or not) at the same time
+        if (trie.left.isEmpty() == trie.right.isEmpty()) {
             return trie;
         }
 
         Trie child;
         byte childImplicitByte;
-        if (leftOpt.isPresent()) {
-            child = leftOpt.get();
+        if (!trie.left.isEmpty()) {
+            child = trie.left.getNode().orElse(null);
             childImplicitByte = (byte) 0;
         } else { // has right node
-            child = rightOpt.get();
+            child = trie.right.getNode().orElse(null);
             childImplicitByte = (byte) 1;
         }
 
+        // could not retrieve from database
+        if (child == null) {
```

Comment:
why did you add this condition? From my understanding, this should never happen.

Reply:
Yep, we think the same, but since this "child" is the result of evaluating an `Optional` field, we have to take that in account. Also note that the code behavior before this change was exactly the same, so this is also for the sake of maintaining it the same.

Reply:
mmm isn't this an unnecesary check? if you do the `Trie.getNode()` you can get the broken db error

```java
public Optional<Trie> getNode() {
        if (lazyNode != null) {
            return Optional.of(lazyNode);
        }

        if (lazyHash == null) {
            return Optional.empty();
        }

        Optional<Trie> node = store.retrieve(lazyHash.getBytes());

        // Broken database, can't continue
        if (!node.isPresent()) {
            logger.error("Broken database, execution can't continue");
            nodeStopper.stop(1);
            return Optional.empty();
        }

        lazyNode = node.get();

        return node;
    }
```